snuba/query/validation/validators.py

```diff
@@ -55,6 +71,38 @@ def validate(self, query: Query, alias: Optional[str] = None) -> None:
             )
 
 
+class EntityContainsColumnsValidator(QueryValidator):
+    """
+    Ensures that all columns in the query actually exist in the entity.
+    """
+
+    def __init__(
+        self, entity_data_model: ColumnSet, validation_mode: ColumnValidationMode
+    ) -> None:
+        self.validation_mode = validation_mode
+        self.entity_data_model = entity_data_model
+
+    def validate(self, query: Query, alias: Optional[str] = None) -> None:
+        if self.validation_mode == ColumnValidationMode.DO_NOTHING:
+            return
+
+        query_columns = query.get_all_ast_referenced_columns()
+
+        missing = set()
+        for column in query_columns:
+            if column.table_name == alias and column.column_name not in {
+                c.name for c in self.entity_data_model.columns
```

Comment:
Can you move `{c.name for c in self.entity_data_model.columns}` outside the loop so we don't have an n^2 thing going on here. Also I think columnsets already implement a `__contains__` function, is there a reason you're not leveraging that?